### PR TITLE
Hints, Problems and Flag functions

### PR DESCRIPTION
**Pull request by KreativeThinker**
Adds functionality for:
- [x] Flag checking
- [x] Hints
- [x] Problems

`get_team_id()` needs to be modified to retrieve the team ID from cookies.
` flag_get` For testing purposes only. flag to be passed in the body of `POST` function. Use commented function

### CODE REVIEW

**Comment by KreativeThinker** [comment]
Added endpoint for team/list. Error in creating a user which requires further examination. Awaiting user authentication and the addition of point management to add the leaderboard and the team creation functionality.

File: src/pwncore/routes/team.py [hunk 1]
```diff
@@ -9,9 +11,12 @@
 
 
 @router.get("/list")
-async def team_list():
-    # Do login verification here
-    return [{"team_name": "CID Squad"}, {"team_name": "Astra"}]
+async def team_list(response: Response):
+    teams = await Team.all().values()
+    if not teams:
```
Comment (mradigen) [inline src/pwncore/routes/team.py:16]: No need to check for teams being empty, if there are no teams return the empty list.

File: src/pwncore/routes/team.py [hunk 2]
```diff
@@ -20,7 +25,12 @@ async def team_login():
     return {"status": "logged in!"}
 
 
+# Unable to test as adding users returns an error
```
Comment (mradigen) [inline src/pwncore/routes/team.py:28]: The model `User.save` method has a bug: It is not checking whether `members` fields exists on the Team before trying to filter for it.

In `src/pwncore/models/user.py`, change:
```py
- if self.team is not None:
+ if self.team is not None and hasattr(self.team, "members"):
```

File: src/pwncore/routes/__init__.py [hunk 3]
```diff
@@ -1,10 +1,67 @@
...
+            image_name="key:latest",
+            image_config={"PortBindings": {"22/tcp": [{}]}},
+        )
+        await Team.create(
+            name="CID Squad" + uuid.uuid4().hex, secret_hash="veryverysecret"
+        )
+        await Team.create(
+            name="Triple A battery" + uuid.uuid4().hex, secret_hash="chotiwali"
+        )
+        await User.create(name="Aditya Jyoti", team_id=2)
```
Comment (mradigen) [inline src/pwncore/routes/__init__.py:37]: Given the User model is fixed, you'll have to provide the tag (reg. no.), email, and phone number for the user to be created.

File: src/pwncore/routes/team.py [hunk 4]
```diff
@@ -20,7 +25,12 @@ async def team_login():
...
 @router.get("/members/{team_id}")
-async def team_members(team_id: int):
-    # Get team members from team_id
-    return [{"name": "ABC", "user_id": 3432}, {"name": "DEF", "user_id": 3422}]
+async def team_members(team_id: int, response: Response):
+    team = await Team.get_or_none(id=team_id).values_list("members_id", flat=True)
+    if not team:
+        response.status_code = 404
+        return {"msg_code": config.msg_codes["team_not_found"]}
+    members = [await User.get(id=x) for x in team]
```
Comment (mradigen) [inline src/pwncore/routes/team.py:35]: Filter directly for team_id in User instead:
```py
members = await User.filter(team_id=team_id).values()
```

**Review by mradigen — APPROVED** [review]

File: src/pwncore/models/ctf.py [hunk 5]
```diff
@@ -48,7 +48,9 @@ class Meta:
 
 class ViewedHint(Model):
     team: fields.ForeignKeyRelation[Team] = fields.ForeignKeyField("models.Team")
-    hint: fields.ForeignKeyRelation[Hint] = fields.ForeignKeyField("models.Hint")
+    hint: fields.ForeignKeyRelation[Hint] = fields.ForeignKeyField(
+        "models.Hint", related_name="hints"
```
Comment (WizzyGeek) [inline src/pwncore/models/ctf.py:52]: why is related_name added here?
Why is Hint.hints pointing to ViewedHint records?
Remove if not intended

File: src/pwncore/models/user.py [hunk 6]
```diff
@@ -29,7 +29,7 @@ class User(Model):
     async def save(self, *args, **kwargs):
         # TODO: Insert/Update in one query
         # Reason why we dont use pre_save: overhead, ugly
-        if self.team is not None:
+        if self.team is not None and hasattr(self.team, "members"):
```
[the remark this answers is not in the document]
Comment (WizzyGeek) [reply src/pwncore/models/user.py:32]: Give me a minimal repro if possible btw

**Review by WizzyGeek — CHANGES_REQUESTED** [review]

File: src/pwncore/models/user.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (KreativeThinker) [reply src/pwncore/models/user.py:32]: https://github.com/lugvitc/pwncore/pull/10#discussion_r1398646209

There was an error creating new user s which was resolved by adding that statement.

**Comment by KreativeThinker** [comment]
- [x] Changed routes.
- [x] Switched to pydantic serialization for outputs.
- [x] Switched to .exists
- [x] Removed function to view all viewed hints (required?)
- [x] Added error code `{ctf_solved: 12}`
- [ ] Remove init_db
- [ ] Rewrite `hint_get`
- [ ] Refactoring

File: src/pwncore/models/ctf.py [hunk 7]
```diff
@@ -38,7 +43,7 @@ class Meta:
 class SolvedProblem(Model):
     team: fields.ForeignKeyRelation[Team] = fields.ForeignKeyField("models.Team")
     problem: fields.ForeignKeyRelation[Problem] = fields.ForeignKeyField(
-        "models.Problem"
+        "models.Problem", related_name="problem"
```
Comment (WizzyGeek) [inline src/pwncore/models/ctf.py:46]: related_name shouldn't be here

**Review by WizzyGeek — CHANGES_REQUESTED** [review]

**Comment by WizzyGeek** [comment]
What happened with the git history here?

Well anyway code looks good, nice work @KreativeThinker and @mradigen

**Comment by KreativeThinker** [comment]
Need to make some changes for consistency. Also, @mradigen has recommended using custom json outputs. Should I switch to the recommended model or proceed with the current pydantic model?

**Comment by mradigen** [comment]
> Need to make some changes for consistency. Also, @mradigen has recommended using custom json outputs. Should I switch to the recommended model or proceed with the current pydantic model?

Feel free to disregard my suggestion, and if it's convenient for you, proceed with the Pydantic model. I realised the benefit is that it allows the docs to know exactly what the routes will return, simplifying frontend development.

File: src/pwncore/models/container.py [hunk 8]
```diff
@@ -18,9 +18,7 @@ class Container(Model):
     problem: fields.ForeignKeyRelation[Problem] = fields.ForeignKeyField(
         "models.Problem", on_delete=fields.OnDelete.NO_ACTION
     )
-    team: fields.ForeignKeyRelation[Team] = fields.ForeignKeyField(
-        "models.Team", related_name="containers"
```
Comment (WizzyGeek) [inline src/pwncore/models/container.py:22]: Readd the related_name or remove the `containers` annotation in the Team class

File: src/pwncore/models/ctf.py [hunk 9]
```diff
@@ -19,15 +23,19 @@ class Problem(Model):
...
+    class PydanticMeta:
+        exclude = ["image_name", "image_config"]
+
 
 class Hint(Model):
+    id = fields.IntField(pk=True)
     order = fields.SmallIntField()  # 0, 1, 2
     problem: fields.ForeignKeyRelation[Problem] = fields.ForeignKeyField(
-        "models.Problem", related_name="hints"
+        "models.Problem"
```
Comment (WizzyGeek) [inline src/pwncore/models/ctf.py:38]: Readd the `related_name`

File: src/pwncore/routes/ctf/__init__.py [hunk 10]
```diff
@@ -10,19 +24,89 @@
...
+
+# For testing purposes only. flag to be passed in body of POST function.
+@router.get("/{ctf_id}/flag")
+async def flag_get(ctf_id: int, response: Response):
+    problem = await Problem.exists(id=ctf_id)
+    if not problem:
+        response.status_code = 404
+        return {"msg_code": config.msg_codes["ctf_not_found"]}
+
+    status = await SolvedProblem.get_or_none(team_id=get_team_id(), problem_id=ctf_id)
```
Comment (WizzyGeek) [inline src/pwncore/routes/ctf/__init__.py:50]: .exists

File: src/pwncore/routes/ctf/__init__.py [hunk 11]
```diff
@@ -10,19 +24,89 @@
...
+    if status:
+        response.status_code = 401
+        return {"msg_code": config.msg_codes["ctf_solved"]}
+
+    check_solved = await Container.get_or_none(
```
Comment (WizzyGeek) [inline src/pwncore/routes/ctf/__init__.py:55]: .exists

File: src/pwncore/routes/__init__.py [hunk 12]
```diff
@@ -1,16 +1,11 @@
 from fastapi import APIRouter
 
 from pwncore.routes import ctf, team
+from pwncore.config import config
```
Comment (WizzyGeek) [inline src/pwncore/routes/__init__.py]: Unused import?

File: src/pwncore/routes/ctf/__init__.py [hunk 13]
```diff
@@ -10,19 +24,89 @@
...
+    viewed_hints = await Hint_Pydantic.from_queryset(
+        Hint.filter(problem_id=ctf_id, viewedhints__team_id=get_team_id())
+    )
+    return viewed_hints
+
+
+@router.get("/completed")
+async def completed_problem_get():
+    problems = await Problem_Pydantic.from_queryset(
+        Problem.filter(solvedproblems__team_id=get_team_id())
```
Comment (WizzyGeek) [inline src/pwncore/routes/ctf/__init__.py:101]: `solvedproblems__team_id` does this run? 
change it to `solvedproblem__team_id`

File: src/pwncore/routes/ctf/__init__.py [hunk 14]
```diff
@@ -10,19 +24,89 @@
...
+        hint = await Hint.get(problem_id=ctf_id, order=0)
+
+    await ViewedHint.create(hint_id=hint.id, team_id=get_team_id())
+    return {"text": hint.text, "order": hint.order}
+
+
+@router.get("/{ctf_id}/viewed_hints")
+async def viewed_problem_hints_get(ctf_id: int):
+    viewed_hints = await Hint_Pydantic.from_queryset(
+        Hint.filter(problem_id=ctf_id, viewedhints__team_id=get_team_id())
```
Comment (WizzyGeek) [inline src/pwncore/routes/ctf/__init__.py:93]: `viewedhints__team_id` -> `viewedhint__teami_id`

Comment (WizzyGeek) [reply src/pwncore/routes/ctf/__init__.py:93]: or you can create and explicit relation and set `related_name` to whatever you want

File: src/pwncore/models/ctf.py (hunk 9 above)
Comment (KreativeThinker) [reply src/pwncore/models/ctf.py:38]: We are not using it right now. Should I still readd it?

File: src/pwncore/routes/ctf/__init__.py (hunk 13 above)
Comment (KreativeThinker) [reply src/pwncore/routes/ctf/__init__.py:101]: solvedproblems is generated by tortoise automatically. It runs. I will have to read as related_name field.

File: src/pwncore/models/ctf.py (hunk 9 above)
Comment (WizzyGeek) [reply src/pwncore/models/ctf.py:38]: I remember some code using hints to prefetch, if it was in this PR only then get rid of its annotation in Problem

File: src/pwncore/routes/ctf/__init__.py (hunk 13 above)
Comment (WizzyGeek) [reply src/pwncore/routes/ctf/__init__.py:101]: > solvedproblems is generated by tortoise automatically. It runs. I will have to read as related_name field.

Btw, where is this documented?

Comment (KreativeThinker) [reply src/pwncore/routes/ctf/__init__.py:101]: It is not very well documented but the autogenerated name comes up in the errors when running without a related_name and using an inaccessible key. I can change it in the next commit if needed.

File: src/pwncore/models/ctf.py (hunk 9 above)
Comment (KreativeThinker) [reply src/pwncore/models/ctf.py:38]: All prefetch queries were replaced with automatic accessing using Model.field syntax.

File: src/pwncore/routes/ctf/__init__.py (hunk 13 above)
Comment (WizzyGeek) [reply src/pwncore/routes/ctf/__init__.py:101]: no need, its fine

File: src/pwncore/routes/ctf/__init__.py (hunk 14 above)
Comment (WizzyGeek) [reply src/pwncore/routes/ctf/__init__.py:93]: doesn't need to changed either, this is at your discretion

File: src/pwncore/models/ctf.py (hunk 9 above)
Comment (WizzyGeek) [reply src/pwncore/models/ctf.py:38]: then just remove the ReverseRelation annotation in the Problem model

**Review by mradigen — APPROVED** [review]

**Review by WizzyGeek — APPROVED** [review]
LGTM

**Comment by mradigen** [comment]
Merging this before #6 since resolving potential conflicts with #6 is easier given it has only 8 commits.

**Pull request merged**